### ISSUE TITLE
bug decomposing reified global constraints that contain other global constraints

### ISSUE THREAD

**Issue by tias** [CPMpy/cpmpy#205]
```python
lhs = boolvar(name="lhs")
rhs = boolvar(3, name="rhs")
e = lhs.implies(Xor(rhs[:3]))

print("e:", e)
s = SolverLookup.get("ortools")
#print("trf:", s.transform(e))  # if you are on the #204 branch
s += [e, ~lhs]
s.solve()
```

This reified XOR case is a stinky case that is highly relevant to the refactor in #204 ...

- `reify_rewrite()` of bv.implies(Xor(boolvar(3))) will decompose the Xor, then add impliciations to it.
- it will then flatten this
- however, the decompose of Xor with len>2 will contain new Xors. So the implications added to it should also be `reify_rewrite()`n; in fact any transformation upto and including reify_rewrite should be applied. (or in the most conservative sense: the entire transform function should be applied to it).

This reveals an unanswered design question: who should decompose globals? Should the new `transform()` in #204 do it or should `_post_constraint()` do it? Currently it is the latter...

- Con: this makes debugging harder... when outputting transform it looks as if the solver will support it
- Pro: it does not require transform to know which global constraints are supported
- Con: it is inconsistently implemented, because reified Boolean globals are decomposed in reify_rewrite
- Pro: it allows to apply all transformations on the decomposition again

So either we decompose all globals in transform (and give the transform() function to some of the transformations, possibly doing some double work)

or we don't decompose globals for reification either and then the solver interfaces require two more cases that check for reified globals (once in ->, once in ==)

Currently `reify_rewrite` is only used by ortools/gurobi, not by pysat and pysdd. So for those we also need to decide to either add reify_rewrite to the transformation chain and have that decompose reified globals, or to add the logic to the solver interface indeed.

Not sure...

**Comment by tias** [comment]
The other place is `negated_normal()`, that one also decomposes a global constraint and negates that.

Because it is in flatten, and we can safely assume that flatten is called as first transform, this is generally OK.

However, reify_rewrite/to_cnf/linearize also use that function, when splitting BE == BV into BV -> BE and BE -> BV :: ~BV -> ~BE. It is in this case that the decompose can contain other globals or expressions that need to go through earlier transformations too again...

so if we only decompose globals in the solverinterface, we should keep the BE -> BVs till there too?

**Comment by tias** [comment]
This is being properly addressed with a tree-based decomposer in #335
